engine/collection/compliance/core.go

```diff
@@ -319,3 +319,15 @@ func (c *Core) prunePendingCache() {
 	// always record the metric
 	c.mempoolMetrics.MempoolEntries(metrics.ResourceClusterProposal, c.pending.Size())
 }
+
+// ProcessFinalizedBlock performs pruning of stale data based on finalization event
+// removes pending blocks below the finalized height and prunes vote aggregator
+func (c *Core) ProcessFinalizedBlock(final *flow.Header) {
+	// remove all pending blocks at or below the finalized height
+	c.pending.PruneByHeight(final.Height)
+
+	// always record the metric
+	c.mempoolMetrics.MempoolEntries(metrics.ResourceClusterProposal, c.pending.Size())
+
+	c.voteAggregator.PruneUpToView(final.View)
```

Comment:
Now we have a proper fix than the hotfix to the original issue. Do we also want to add some logs for debugging purpose. The pending mempool size is recorded with metrics, [maybe we could add a debug level log for vote aggregator vote collectors size](https://github.com/onflow/flow-go/pull/2016/files)?

Reply:
I can add a log, but we can also make a metric for vote aggregator size, maybe that will be more informative, most of the time it should be low, but if finalization gets delayed than we will see spikes.

Reply:
yeah, both work for me.

Reply:
@zhangchiqing https://github.com/onflow/flow-go/pull/2040/commits/e5b277f1b7f75a323465bd9d64942631e208a69c